File: Sources/KelvinShift/ScheduleEngine.swift
// KelvinShift – ScheduleEngine.swift
//
// Runs a 15-second timer that:
//   1. Determines the current phase (day / night / transitioning).
//   2. Calculates the target Kelvin for this instant.
//   3. Applies it via direct gamma table manipulation.
//   4. Publishes state for the status-bar UI.
//
// Also supports a "preview" mode where an external caller (the Preferences
// sliders) can temporarily override the display Kelvin while the user drags.

import Foundation

// MARK: – Public State

enum SchedulePhase: String {
    case day, night, transitionToNight, transitionToDay
}

struct ScheduleState {
    let phase: SchedulePhase
    let currentKelvin: Int
    let currentBrightness: Double
    let dayKelvin: Int
    let nightKelvin: Int
    let dayBrightness: Double
    let nightBrightness: Double
    let sunriseTime: Date?
    let sunsetTime: Date?
    let nextEvent: Date?
    let enabled: Bool
}

// MARK: – Engine

final class ScheduleEngine {

    static let stateDidChange = Notification.Name("KSStateChanged")

    /// Convenient shared reference so PreferencesView can call preview methods.
    static weak var current: ScheduleEngine?

    private let gamma = GammaController.shared
    private let settings = Settings.shared
    private var timer: Timer?

    private(set) var state: ScheduleState

    // MARK: – Preview

    /// When non-nil, the engine skips applying its scheduled Kelvin;
    /// the preview value is applied directly instead.
    private var previewKelvin: Int?
    private var previewBrightness: Double?

    // MARK: – Transition Demo

    /// Timer for running the transition demo
    private var demoTimer: Timer?
    /// Current progress through the demo (0.0 to 1.0 represents a full day cycle)
    private(set) var demoProgress: Double = 0.0
    /// Whether the demo is currently running
    private(set) var isDemoRunning: Bool = false
    /// Callback for demo progress updates
    var onDemoProgressChanged: ((Double) -> Void)?

    /// Called when the user begins dragging a Kelvin slider.
    func startPreview(_ kelvin: Int) {
        previewKelvin = kelvin
        previewBrightness = nil
        gamma.applyKelvin(kelvin)
    }

    /// Called on every slider value change while dragging.
    func updatePreview(_ kelvin: Int) {
        guard previewKelvin != nil else { return }
        previewKelvin = kelvin
        gamma.applyKelvin(kelvin)
    }

    /// Called when the user begins dragging a brightness slider.
    func startBrightnessPreview(_ brightness: Double) {
        previewBrightness = brightness
        previewKelvin = nil
        // Use current scheduled kelvin with preview brightness
        gamma.applyKelvinWithBrightness(state.currentKelvin, brightness: brightness)
    }

    /// Called on every brightness slider value change while dragging.
    func updateBrightnessPreview(_ brightness: Double) {
        guard previewBrightness != nil else { return }
        previewBrightness = brightness
        gamma.applyKelvinWithBrightness(state.currentKelvin, brightness: brightness)
    }

    /// Called when the user releases the slider. The engine immediately
    /// re-applies the schedule's correct Kelvin.
    func stopPreview() {
        previewKelvin = nil
        previewBrightness = nil
        restoreScheduledSettings()
    }

    /// Restores the scheduled Kelvin and brightness based on current time
    private func restoreScheduledSettings() {
        let now = Date()
        let (dayMin, nightMin, _, _) = scheduleTimes(for: now)
        let nowMin = minutesFromMidnight(now)
        let tranMin = settings.transitionMinutes
        let nightTransStart = wrap(nightMin - tranMin)
        let dayTransStart = wrap(dayMin - tranMin)

        let kelvin: Int
        let brightness: Double
        if inRange(nowMin, from: dayMin, to: nightTransStart) {
            kelvin = settings.dayKelvin
            brightness = settings.dayBrightness
        } else if inRange(nowMin, from: nightTransStart, to: nightMin) {
            let p = progress(nowMin, from: nightTransStart, length: tranMin)
            kelvin = lerp(settings.dayKelvin, settings.nightKelvin, p)
            brightness = lerpD(settings.dayBrightness, settings.nightBrightness, p)
        } else if inRange(nowMin, from: nightMin, to: dayTransStart) {
            kelvin = settings.nightKelvin
            brightness = settings.nightBrightness
        } else {
            let p = progress(nowMin, from: dayTransStart, length: tranMin)
            kelvin = lerp(settings.nightKelvin, settings.dayKelvin, p)
            brightness = lerpD(settings.nightBrightness, settings.dayBrightness, p)
        }

        gamma.applyKelvinWithBrightness(kelvin, brightness: brightness)
    }

    /// Starts the transition demo, cycling through a full day in the specified duration
    /// - Parameter durationSeconds: Total duration for one full day cycle (default: 10 seconds)
    func startDemo(durationSeconds: Double = 10.0) {
        guard !isDemoRunning else { return }
        isDemoRunning = true
        demoProgress = 0.0

        // Cancel any slider preview
        previewKelvin = nil
        previewBrightness = nil

        let updateInterval = 1.0 / 60.0 // 60 fps for smooth animation
        let progressPerTick = updateInterval / durationSeconds

        demoTimer = Timer.scheduledTimer(withTimeInterval: updateInterval, repeats: true) { [weak self] _ in
            guard let self = self else { return }

            self.demoProgress += progressPerTick
            if self.demoProgress >= 1.0 {
                self.stopDemo()
                return
            }

            self.applyDemoSettings(at: self.demoProgress)
            self.onDemoProgressChanged?(self.demoProgress)
        }
        RunLoop.current.add(demoTimer!, forMode: .common)

        // Apply initial settings
        applyDemoSettings(at: 0.0)
        onDemoProgressChanged?(0.0)
    }

    /// Stops the transition demo and restores scheduled settings
    func stopDemo() {
        demoTimer?.invalidate()
        demoTimer = nil
        isDemoRunning = false
        demoProgress = 0.0
        onDemoProgressChanged?(0.0)
        restoreScheduledSettings()
    }

    /// Applies color temperature and brightness for a given position in the demo cycle
    /// - Parameter demoProgress: Progress through demo (0.0 = day, 0.5 = night, 1.0 = back to day)
    private func applyDemoSettings(at demoProgress: Double) {
        // Simple cycle: day → night → day
        // First half (0.0 to 0.5): transition from day to night
        // Second half (0.5 to 1.0): transition from night to day

        let kelvin: Int
        let brightness: Double

        if demoProgress < 0.5 {
            // Transitioning from day to night (0.0 → 0.5 maps to 0.0 → 1.0)
            let p = demoProgress * 2.0
            kelvin = lerp(settings.dayKelvin, settings.nightKelvin, p)
            brightness = lerpD(settings.dayBrightness, settings.nightBrightness, p)
        } else {
            // Transitioning from night to day (0.5 → 1.0 maps to 0.0 → 1.0)
            let p = (demoProgress - 0.5) * 2.0
            kelvin = lerp(settings.nightKelvin, settings.dayKelvin, p)
            brightness = lerpD(settings.nightBrightness, settings.dayBrightness, p)
        }

        gamma.applyKelvinWithBrightness(kelvin, brightness: brightness)
    }

    // MARK: – Init

    init() {
        self.state = ScheduleState(
            phase: .day, currentKelvin: 6500, currentBrightness: 1.0,
            dayKelvin: 5000, nightKelvin: 2700,
            dayBrightness: 1.0, nightBrightness: 0.8,
            sunriseTime: nil, sunsetTime: nil,
            nextEvent: nil, enabled: true
        )
        Self.current = self

        NotificationCenter.default.addObserver(
            self, selector: #selector(onSettingsChanged),
            name: Settings.didChange, object: nil
        )
    }

    func start() {
        tick()                     // apply immediately
        timer = Timer.scheduledTimer(withTimeInterval: 15, repeats: true) { [weak self] _ in
            self?.tick()
        }
        RunLoop.current.add(timer!, forMode: .common)
    }

    func stop() {
        timer?.invalidate()
        timer = nil
        gamma.resetGamma()
    }

    // MARK: – Timer callback

    @objc private func onSettingsChanged() {
        guard previewKelvin == nil && previewBrightness == nil && !isDemoRunning else { return }
        tick()
    }

    private func tick() {
        guard previewKelvin == nil && previewBrightness == nil && !isDemoRunning else { return }

        guard settings.enabled else {
            gamma.resetGamma()
            publish(.day, kelvin: 6500, brightness: 1.0, sunrise: nil, sunset: nil, next: nil, enabled: false)
            return
        }

        let now = Date()
        let (dayMin, nightMin, sunrise, sunset) = scheduleTimes(for: now)
        let nowMin   = minutesFromMidnight(now)
        let tranMin  = settings.transitionMinutes

        let nightTransStart = wrap(nightMin - tranMin)
        let dayTransStart   = wrap(dayMin   - tranMin)

        let kelvin: Int
        let brightness: Double
        let phase: SchedulePhase
        var next: Date? = nil

        if inRange(nowMin, from: dayMin, to: nightTransStart) {
            phase  = .day
            kelvin = settings.dayKelvin
            brightness = settings.dayBrightness
            next   = todayAt(nightTransStart, relativeTo: now)

        } else if inRange(nowMin, from: nightTransStart, to: nightMin) {
            phase  = .transitionToNight
            let p  = progress(nowMin, from: nightTransStart, length: tranMin)
            kelvin = lerp(settings.dayKelvin, settings.nightKelvin, p)
            brightness = lerpD(settings.dayBrightness, settings.nightBrightness, p)
            next   = todayAt(nightMin, relativeTo: now)

        } else if inRange(nowMin, from: nightMin, to: dayTransStart) {
            phase  = .night
            kelvin = settings.nightKelvin
            brightness = settings.nightBrightness
            next   = todayAt(dayTransStart, relativeTo: now)

        } else {
            phase  = .transitionToDay
            let p  = progress(nowMin, from: dayTransStart, length: tranMin)
            kelvin = lerp(settings.nightKelvin, settings.dayKelvin, p)
            brightness = lerpD(settings.nightBrightness, settings.dayBrightness, p)
            next   = todayAt(dayMin, relativeTo: now)
        }

        gamma.applyKelvinWithBrightness(kelvin, brightness: brightness)
        publish(phase, kelvin: kelvin, brightness: brightness, sunrise: sunrise, sunset: sunset, next: next, enabled: true)
    }

    // MARK: – Schedule helpers

    private func scheduleTimes(for date: Date) -> (Int, Int, Date?, Date?) {
        if settings.scheduleMode == "solar",
           let s = SolarCalculator.calculate(date: date,
                                             latitude: settings.latitude,
                                             longitude: settings.longitude) {
            return (minutesFromMidnight(s.sunrise),
                    minutesFromMidnight(s.sunset),
                    s.sunrise, s.sunset)
        }
        return (settings.customDayHour   * 60 + settings.customDayMinute,
                settings.customNightHour * 60 + settings.customNightMinute,
                nil, nil)
    }

    // MARK: – Circular-clock math

    private func wrap(_ m: Int) -> Int { ((m % 1440) + 1440) % 1440 }

    private func inRange(_ t: Int, from: Int, to: Int) -> Bool {
        from <= to
            ? (t >= from && t < to)
            : (t >= from || t < to)
    }

    private func progress(_ t: Int, from: Int, length: Int) -> Double {
        let elapsed = wrap(t - from)
        return min(1, Double(elapsed) / Double(max(1, length)))
    }

    private func lerp(_ a: Int, _ b: Int, _ t: Double) -> Int {
        let s = t * t * (3 - 2 * t)
        return a + Int((Double(b - a) * s).rounded())
    }

    private func lerpD(_ a: Double, _ b: Double, _ t: Double) -> Double {
        let s = t * t * (3 - 2 * t)
        return a + (b - a) * s
    }

    private func minutesFromMidnight(_ date: Date) -> Int {
        let c = Calendar.current
        return c.component(.hour, from: date) * 60 + c.component(.minute, from: date)
    }

    private func todayAt(_ minutes: Int, relativeTo ref: Date) -> Date {
        let startOfDay = Calendar.current.startOfDay(for: ref)
        var d = startOfDay.addingTimeInterval(Double(minutes) * 60)
        if d < ref { d = d.addingTimeInterval(86400) }
        return d
    }

    // MARK: – State publishing

    private func publish(
        _ phase: SchedulePhase, kelvin: Int, brightness: Double,
        sunrise: Date?, sunset: Date?, next: Date?, enabled: Bool
    ) {
        state = ScheduleState(
            phase: phase, currentKelvin: kelvin, currentBrightness: brightness,
            dayKelvin: settings.dayKelvin, nightKelvin: settings.nightKelvin,
            dayBrightness: settings.dayBrightness, nightBrightness: settings.nightBrightness,
            sunriseTime: sunrise, sunsetTime: sunset,
            nextEvent: next, enabled: enabled
        )
        NotificationCenter.default.post(name: Self.stateDidChange, object: state)
    }
}

File: win/src/ScheduleEngine.cpp
#include "ScheduleEngine.h"
#include "SettingsService.h"
#include "GammaService.h"
#include "SolarCalculator.h"
#include <algorithm>
#include <cmath>

// ── Circular-clock helpers ────────────────────────────────────────────────
namespace {

int Wrap(int m) { return ((m % 1440) + 1440) % 1440; }

bool InRange(int t, int from, int to)
{
    return from <= to ? (t >= from && t < to)
                      : (t >= from || t < to);
}

double Progress(int t, int from, int length)
{
    int elapsed = Wrap(t - from);
    return std::min(1.0, (double)elapsed / std::max(1, length));
}

// Linear interpolation — constant rate of change so a fixed slice of time
// always moves the same colour delta regardless of where in the ramp it is.
int LerpI(int a, int b, double t) { return a + (int)llround((b - a) * t); }
double LerpD(double a, double b, double t) { return a + (b - a) * t; }

int MinutesFromMidnight(const SYSTEMTIME& d) { return d.wHour * 60 + d.wMinute; }

// Normalize raw solar minutes (possibly <0 or >=1440) into a time-of-day,
// matching DateTime.AddMinutes(...).Hour*60+Minute in the C# port.
int NormalizeMinute(double x) { return Wrap((int)std::floor(x)); }

} // namespace

ScheduleEngine::ScheduleEngine(SettingsService& settings, GammaService& gamma)
    : settings_(settings), gamma_(gamma)
{
    state_.dayKelvin = settings_.DayKelvin();
    state_.nightKelvin = settings_.NightKelvin();
    state_.dayBrightness = settings_.DayBrightness();
    state_.nightBrightness = settings_.NightBrightness();
    state_.bedtimeEnabled = settings_.BedtimeEnabled();
    state_.bedtimeKelvin = settings_.BedtimeKelvin();
    state_.bedtimeBrightness = settings_.BedtimeBrightness();
    state_.enabled = settings_.Enabled();

    timer_.SetInterval(15000); // 15-second schedule tick
    timer_.SetCallback([this] { Tick(); });

    settings_.SettingsChanged.Add([this]
    {
        if (!previewKelvin_ && !previewBrightness_ && !isDemoRunning_)
            Tick();
    });
}

void ScheduleEngine::Start()
{
    Tick();
    timer_.Start();
}

void ScheduleEngine::Stop()
{
    timer_.Stop();
    gamma_.Reset();
}

// ── Preview ───────────────────────────────────────────────────────────────

void ScheduleEngine::StartPreview(int kelvin)
{
    previewKelvin_ = kelvin;
    previewBrightness_.reset();
    gamma_.ApplyKelvin(kelvin);
}

void ScheduleEngine::UpdatePreview(int kelvin)
{
    if (!previewKelvin_) return;
    previewKelvin_ = kelvin;
    gamma_.ApplyKelvin(kelvin);
}

void ScheduleEngine::StartBrightnessPreview(double brightness)
{
    previewBrightness_ = brightness;
    previewKelvin_.reset();
    gamma_.ApplyKelvinWithBrightness(state_.currentKelvin, brightness);
}

void ScheduleEngine::UpdateBrightnessPreview(double brightness)
{
    if (!previewBrightness_) return;
    previewBrightness_ = brightness;
    gamma_.ApplyKelvinWithBrightness(state_.currentKelvin, brightness);
}

void ScheduleEngine::StopPreview()
{
    previewKelvin_.reset();
    previewBrightness_.reset();
    RestoreScheduledSettings();
}

void ScheduleEngine::RestoreScheduledSettings()
{
    SYSTEMTIME now;
    GetLocalTime(&now);
    auto r = ComputeSchedule(now);
    gamma_.ApplyKelvinWithBrightness(r.kelvin, r.brightness);
}

// ── Demo cycle ────────────────────────────────────────────────────────────

void ScheduleEngine::StartDemo(double durationSeconds)
{
    if (isDemoRunning_) return;
    isDemoRunning_ = true;
    demoProgress_ = 0.0;
    previewKelvin_.reset();
    previewBrightness_.reset();

    const double step = 1.0 / 60.0;
    demoPerTick_ = step / durationSeconds;

    demoTimer_.SetInterval((unsigned)std::lround(step * 1000.0));
    demoTimer_.SetCallback([this]
    {
        demoProgress_ += demoPerTick_;
        if (demoProgress_ >= 1.0) { StopDemo(); return; }
        ApplyDemoSettings(demoProgress_);
        DemoProgressChanged.Raise(demoProgress_);
    });
    demoTimer_.Start();
    ApplyDemoSettings(0.0);
    DemoProgressChanged.Raise(0.0);
}

void ScheduleEngine::StopDemo()
{
    demoTimer_.Stop();
    isDemoRunning_ = false;
    demoProgress_ = 0.0;
    DemoProgressChanged.Raise(0.0);
    RestoreScheduledSettings();
}

void ScheduleEngine::ApplyDemoSettings(double p)
{
    // "Preview Cycle" always runs the full day -> night -> bedtime -> day
    // loop, so the bedtime phase is always shown — the point of the preview
    // is to see it, even before bedtime is switched on in the schedule.
    int k;
    double b;
    const double third = 1.0 / 3.0;
    if (p < third)
    {
        double t = p * 3.0;
        k = LerpI(settings_.DayKelvin(), settings_.NightKelvin(), t);
        b = LerpD(settings_.DayBrightness(), settings_.NightBrightness(), t);
    }
    else if (p < 2 * third)
    {
        double t = (p - third) * 3.0;
        k = LerpI(settings_.NightKelvin(), settings_.BedtimeKelvin(), t);
        b = LerpD(settings_.NightBrightness(), settings_.BedtimeBrightness(), t);
    }
    else
    {
        double t = (p - 2 * third) * 3.0;
        k = LerpI(settings_.BedtimeKelvin(), settings_.DayKelvin(), t);
        b = LerpD(settings_.BedtimeBrightness(), settings_.DayBrightness(), t);
    }
    gamma_.ApplyKelvinWithBrightness(k, b);
}

// ── Tick ──────────────────────────────────────────────────────────────────

void ScheduleEngine::Tick()
{
    if (previewKelvin_ || previewBrightness_ || isDemoRunning_) return;

    if (!settings_.Enabled())
    {
        gamma_.Reset();
        Publish(SchedulePhase::Day, 6500, 1.0,
                std::nullopt, std::nullopt, std::nullopt, false);
        return;
    }

    SYSTEMTIME now;
    GetLocalTime(&now);
    Times t = ScheduleTimes(now);
    auto r = ComputeSchedule(now);

    gamma_.ApplyKelvinWithBrightness(r.kelvin, r.brightness);
    Publish(r.phase, r.kelvin, r.brightness, t.sunriseMin, t.sunsetMin,
            r.nextMinute, true);
}

// ── Shared computation ────────────────────────────────────────────────────

ScheduleEngine::ScheduleResult ScheduleEngine::ComputeSchedule(const SYSTEMTIME& now) const
{
    // "Day starts" / "Night starts" / "Bedtime" are the times the screen
    // REACHES that state. Each transition runs for tranMin minutes BEFORE
    // the scheduled time and is capped to the gap to the adjacent anchor.
    Times t = ScheduleTimes(now);
    int dayMin = t.dayMin, nightMin = t.nightMin;
    int nowMin = MinutesFromMidnight(now);
    int tranMin = settings_.TransitionMinutes();

    bool useBedtime = BedtimeIsActive(dayMin, nightMin);
    int bedMin = Wrap(settings_.BedtimeHour() * 60 + settings_.BedtimeMinute());
    int morningFromK = useBedtime ? settings_.BedtimeKelvin() : settings_.NightKelvin();
    double morningFromB = useBedtime ? settings_.BedtimeBrightness() : settings_.NightBrightness();

    int dayToNight = Wrap(nightMin - dayMin);
    int nightToBed = useBedtime ? Wrap(bedMin - nightMin) : 0;
    int bedOrNightToDay = useBedtime ? Wrap(dayMin - bedMin)
                                     : Wrap(dayMin - nightMin);

    int nightTransLen = std::min(tranMin, dayToNight);
    int dayTransLen   = std::min(tranMin, bedOrNightToDay);
    int rampLen = useBedtime ? std::min(settings_.BedtimeRampMinutes(), nightToBed) : 0;

    int nightTransStart = Wrap(nightMin - nightTransLen);
    int dayTransStart   = Wrap(dayMin - dayTransLen);

    if (InRange(nowMin, dayMin, nightTransStart))
        return { settings_.DayKelvin(), settings_.DayBrightness(),
                 SchedulePhase::Day, nightTransStart };

    if (InRange(nowMin, nightTransStart, nightMin))
    {
        double p = Progress(nowMin, nightTransStart, nightTransLen);
        return { LerpI(settings_.DayKelvin(), settings_.NightKelvin(), p),
                 LerpD(settings_.DayBrightness(), settings_.NightBrightness(), p),
                 SchedulePhase::TransitionToNight, nightMin };
    }

    if (InRange(nowMin, dayTransStart, dayMin))
    {
        double p = Progress(nowMin, dayTransStart, dayTransLen);
        return { LerpI(morningFromK, settings_.DayKelvin(), p),
                 LerpD(morningFromB, settings_.DayBrightness(), p),
                 SchedulePhase::TransitionToDay, dayMin };
    }

    // Past nightMin, before dayTransStart.
    if (useBedtime)
    {
        int rampStart = Wrap(bedMin - rampLen);
        if (InRange(nowMin, nightMin, rampStart))
            return { settings_.NightKelvin(), settings_.NightBrightness(),
                     SchedulePhase::Night, rampStart };
        if (InRange(nowMin, rampStart, bedMin))
        {
            double p = Progress(nowMin, rampStart, rampLen);
            return { LerpI(settings_.NightKelvin(), settings_.BedtimeKelvin(), p),
                     LerpD(settings_.NightBrightness(), settings_.BedtimeBrightness(), p),
                     SchedulePhase::RampToBedtime, bedMin };
        }
        return { settings_.BedtimeKelvin(), settings_.BedtimeBrightness(),
                 SchedulePhase::Bedtime, dayTransStart };
    }
    return { settings_.NightKelvin(), settings_.NightBrightness(),
             SchedulePhase::Night, dayTransStart };
}

bool ScheduleEngine::BedtimeIsActive(int dayMin, int nightMin) const
{
    if (!settings_.BedtimeEnabled()) return false;
    int bed = Wrap(settings_.BedtimeHour() * 60 + settings_.BedtimeMinute());
    int nightToDay = Wrap(dayMin - nightMin);
    int nightToBed = Wrap(bed - nightMin);
    return nightToBed > 0 && nightToBed < nightToDay;
}

ScheduleEngine::Times ScheduleEngine::ScheduleTimes(const SYSTEMTIME& date) const
{
    if (settings_.ScheduleMode() == "solar")
    {
        auto s = SolarCalculator::Calculate(date, settings_.Latitude(), settings_.Longitude());
        if (s)
        {
            int rise = NormalizeMinute(s->sunriseMin);
            int set  = NormalizeMinute(s->sunsetMin);
            return { rise, set, rise, set };
        }
    }
    return { settings_.CustomDayHour() * 60 + settings_.CustomDayMinute(),
             settings_.CustomNightHour() * 60 + settings_.CustomNightMinute(),
             std::nullopt, std::nullopt };
}

void ScheduleEngine::Publish(SchedulePhase phase, int kelvin, double brightness,
                             std::optional<int> sunrise, std::optional<int> sunset,
                             std::optional<int> next, bool enabled)
{
    state_.phase = phase;
    state_.currentKelvin = kelvin;
    state_.currentBrightness = brightness;
    state_.dayKelvin = settings_.DayKelvin();
    state_.nightKelvin = settings_.NightKelvin();
    state_.dayBrightness = settings_.DayBrightness();
    state_.nightBrightness = settings_.NightBrightness();
    state_.bedtimeEnabled = settings_.BedtimeEnabled();
    state_.bedtimeKelvin = settings_.BedtimeKelvin();
    state_.bedtimeBrightness = settings_.BedtimeBrightness();
    state_.sunriseMin = sunrise;
    state_.sunsetMin = sunset;
    state_.nextEventMin = next;
    state_.enabled = enabled;
    StateChanged.Raise(state_);
}
